Libraries/LibURL/Pattern/PatternParser.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2025, Shannon Booth <shannon@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Function.h>
+#include <LibURL/Pattern/Options.h>
+#include <LibURL/Pattern/Part.h>
+#include <LibURL/Pattern/PatternError.h>
+#include <LibURL/Pattern/Tokenizer.h>
+
+namespace URL::Pattern {
+
+// https://urlpattern.spec.whatwg.org/#pattern-parser
+class PatternParser {
+public:
+    // https://urlpattern.spec.whatwg.org/#encoding-callback
+    // An encoding callback is an abstract algorithm that takes a given string input. The input will be a simple text
+    // piece of a pattern string. An implementing algorithm will validate and encode the input. It must return the
+    // encoded string or throw an exception.
+    using EncodingCallback = Function<PatternErrorOr<String>(String const&)>;
+
+    static PatternErrorOr<Vector<Part>> parse(Utf8View const& input, Options const&, EncodingCallback);
+
+private:
+    PatternParser(EncodingCallback, String segment_wildcard_regexp);
+
+    Token const* try_to_consume_a_token(Token::Type);
+    Token const* try_to_consume_a_modifier_token();
+    Token const* try_to_consume_a_regexp_or_wildcard_token(Token const* name_token);
```

Comment:
hmm I'd be more comfortable if we returned `Optional<Token const&>` here, to avoid any potential UB. But feel free to ignore if that makes handling this type a total pain, as the code does look correct to me.

Reply:
I'm honestly a little iffy/undecided on how I feel about `Optional<T const&>` semantics vs just a pointer.

Though, on UB, project wide, after https://github.com/LadybirdBrowser/ladybird/commit/ef30f191b69dd99b58f39bb732daf86e55fd0f94 I suppose the stance is "yes it's UB, but it will just crash"

Applied the change anyway, it's just changing the type and the pointer checks to `has_value()` instead, and all of this is private to the parser. So super minor. I don't really mind either way.

Reply:
Eh, that's kinda an "if you're not unlucky" thing. I've seen plenty of null derefs in prod happily not crash, so it's nice to not rely on it when it doesn't tank performance.